keps/0031-namespace-management.md

```diff
@@ -0,0 +1,167 @@
+1. namespace comparison could be simple check.  It is unclear when an overwrite should occur or if this should be an error.
+1. Auto create (with a `--apply-ns` or `--ns-ensure`)
+1. Creation of namespace only handled if `namespaceManifest` is provided.  The case of simple namespace (without metadata) would need to be considered.
+
+### Proposal: Multi-Namespace (static)
+
+There is a desire to support multiple namespaces. The desire seems to stem from 2 sources; KUDO early adoptors as operator devs what have large multi-operator needs and those using KUDO for micro-service management. An example of large operator development is Kubeflow, where there is 20+ operators inside D2iQ KUDO Kubeflow operator.  There is a strong desire to logically group many of the supportive infrastructure operators inside their own namespace, reducing the cognitive burden of the end user by removing non-primary operators into a separate namespace from the primary operators.
+ In this model, the expectations are that the 2nd tier namespaces (not the "primary" operator) will have statically defined namespaces which should be honored.   A significant challenge to this is objects with No namespace will land whereever KUDO is configured to install the operator, but other objects will have no control and will be static.   Part of the justification of this is kubeflow, where there are 30+ operators, many which are supportive in nature that the operator developer would like to section off.
+
+This propose is for KUDO to deploy objects to KUDO managed {{.Namespace}} if not specified and to honor `namespace:` for objects that have them.  There are challenges to package validation as we won't know what is accidently missing or accidently specified.  There are challenges to it being static which can be overcome through the use of Params.  It seems if Params are used, that the values for the instance are  immutable after installation.
+
+If we support multi-namespaces in 1 KUDO operator, we need to update package verify to warn on all uses of namespaces that are NOT `{{ .Namespace }}` or empty.
+
+#### Alternative Ideas
+
+1. KUDO to create a cluster scoped instance to own the namespaces
+
+### Proposal: Dependencies and Creating Namespaces
+
+Regarding dependencies and creating namespaces, it is expected that the use of `--namespace` applies to the parent AND all dependent operators AND take the namespace is created prior to applying any dependent steps.  While there is likely desire to have independent control for dependent operators namespaces, this functionality will need to be thoughout more thoroughly regarding dependency management.  This part of the proposal was to provide completeness of namespace management inclusive of the current state of the [KEP-29 Operator Dependencies](0029-operator-dependencies.md) and not to extend further.
```

Comment:
A tricky one but not a blocker: what if a parent and its dependent both have `namespaceManifest` specified and they point to different namespaces?

Reply:
we will only support the creation of 1 namespace even with dependencies for now... anything else needs cluster-wide resource support.  If there is a dependency with 2 different namespaceManifests... we will need to figure out what to do there... it could be a failure... or it could be merged... I would lean towards failure at the moment.

Reply:
> A tricky one but not a blocker: what if a parent and its dependent both have `namespaceManifest` specified and they point to different namespaces?

The top-level `namespace` will be used everywhere, dependencies namespace manifest will be ignored.

Reply:
@zen-dog at this point seems agreeable... I'll update with that detail!